### PR TITLE
[CARBONDATA-4075] Using withEvents instead of fireEvent

### PR DESCRIPTION
**Pull request by QiangCai**
### Why is this PR needed?
 withEvents method can simply code to fire event
 
 ### What changes were proposed in this PR?
Refactor code use the withEvents method instead of fireEvent
    
 ### Does this PR introduce any user interface change?
 - No

 ### Is any new testcase added?
 - No

### CODE REVIEW

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12444/job/ApacheCarbonPRBuilder2.3/5316/

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12444/job/ApacheCarbonPRBuilder2.3/5317/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12444/job/ApacheCarbon_PR_Builder_2.4.5/3557/

**Comment by QiangCai** [comment]
retest this please

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12444/job/ApacheCarbonPRBuilder2.3/5333/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12444/job/ApacheCarbon_PR_Builder_2.4.5/3573/

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/5048/

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/3290/

**Comment by VenuReddy2103** [comment]
We still seem to have few more pre and post events being fired in conventional way. Yet to change them too ?

**Comment by Indhumathi27** [comment]
@QiangCai please rebase

**Comment by QiangCai** [comment]
@Indhumathi27 done

**Comment by QiangCai** [comment]
@VenuReddy2103
 I checked all places, but 14 places still keep the same with previous.
1) in 9 places, it uses only one event, fireEvent is ok. 
2) in 4 places, preEvent and postEvent are in the different code blocks.
3) in 1 place, there is a large code block between preEvent and postEvent.

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12444/job/ApacheCarbonPRBuilder2.3/5522/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12444/job/ApacheCarbon_PR_Builder_2.4.5/3757/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/3342/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/5101/

File: integration/spark/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonAlterTableCompactionCommand.scala [hunk 1]
```diff
@@ -74,7 +74,7 @@ case class CarbonAlterTableCompactionCommand(
       CarbonTable.buildFromTableInfo(tableInfoOp.get)
     } else {
       val relation = CarbonEnv.getInstance(sparkSession).carbonMetaStore
-        .lookupRelation(Option(dbName), tableName)(sparkSession).asInstanceOf[CarbonRelation]
+        .lookupRelation(Option(dbName), tableName)(sparkSession)
```
Comment (Indhumathi27) [inline integration/spark/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonAlterTableCompactionCommand.scala:77]: please remove unused carbonrelation import also

**Comment by Indhumathi27** [comment]
@QiangCai i think you can refactor fireEevents in  SIRebuildSegmentRunner class

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/5053/

**Comment by QiangCai** [comment]
@Indhumathi27 done

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12444/job/ApacheCarbonPRBuilder2.3/5550/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12444/job/ApacheCarbon_PR_Builder_2.4.5/3784/

**Comment by Indhumathi27** [comment]
retest this please

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12444/job/ApacheCarbonPRBuilder2.3/5559/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12444/job/ApacheCarbon_PR_Builder_2.4.5/3793/

**Comment by Indhumathi27** [comment]
LGTM

**Comment by VenuReddy2103** [comment]
LGTM

**Pull request closed without merging**